src/__tests__/type.js

```diff
@@ -600,6 +600,63 @@ test('can type into an input with type `email`', () => {
   expect(element).toHaveValue(email)
 })
 
+test('can type into an input with type `date`', () => {
+  const {element, getEventSnapshot} = setup('<input type="date" />')
+  const date = '2020-06-29'
+  userEvent.type(element, date)
+  expect(element).toHaveValue(date)
+  expect(getEventSnapshot()).toMatchInlineSnapshot(`
+    Events fired on: input[value="2020-06-29"]
+
+    input[value=""] - pointerover
+    input[value=""] - pointerenter
+    input[value=""] - mouseover: Left (0)
+    input[value=""] - mouseenter: Left (0)
+    input[value=""] - pointermove
+    input[value=""] - mousemove: Left (0)
+    input[value=""] - pointerdown
+    input[value=""] - mousedown: Left (0)
+    input[value=""] - focus
+    input[value=""] - focusin
+    input[value=""] - pointerup
+    input[value=""] - mouseup: Left (0)
+    input[value=""] - click: Left (0)
+    input[value=""] - keydown: 2 (50)
+    input[value=""] - keypress: 2 (50)
+    input[value=""] - keyup: 2 (50)
+    input[value=""] - keydown: 0 (48)
+    input[value=""] - keypress: 0 (48)
+    input[value=""] - keyup: 0 (48)
+    input[value=""] - keydown: 2 (50)
+    input[value=""] - keypress: 2 (50)
+    input[value=""] - keyup: 2 (50)
+    input[value=""] - keydown: 0 (48)
+    input[value=""] - keypress: 0 (48)
+    input[value=""] - keyup: 0 (48)
+    input[value=""] - keydown: - (45)
+    input[value=""] - keypress: - (45)
+    input[value=""] - keyup: - (45)
+    input[value=""] - keydown: 0 (48)
+    input[value=""] - keypress: 0 (48)
+    input[value=""] - keyup: 0 (48)
+    input[value=""] - keydown: 6 (54)
+    input[value=""] - keypress: 6 (54)
+    input[value=""] - keyup: 6 (54)
+    input[value=""] - keydown: - (45)
+    input[value=""] - keypress: - (45)
+    input[value=""] - keyup: - (45)
+    input[value=""] - keydown: 2 (50)
+    input[value=""] - keypress: 2 (50)
+    input[value=""] - keyup: 2 (50)
+    input[value=""] - keydown: 9 (57)
+    input[value=""] - keypress: 9 (57)
+    input[value="2020-06-29"] - input
+      "{CURSOR}" -> "{CURSOR}2020-06-29"
```

Comment:
Is this the right spot for the cursor? I would expect it at the end.

Reply:
I hadn't think about that but using it, it seems to keep the last typed part selected. Maybe we could force the selection of the day (the last part of `type('2020-06-29'`))? 

![ezgif com-video-to-gif](https://user-images.githubusercontent.com/2339362/88102389-dbf37880-cb75-11ea-994e-da54be562242.gif)

Reply:
This is where things get complicated... For you the date input accepts `dd/mm/yyyy`, but for me it's `mm/dd/yyyy`. So what the user types is different for everyone. Additionally, here's what happens when I type `10181988`  in the date field in [the playground](https://codesandbox.io/s/user-event-playground-eo909):

```
input#date[value=""] - pointerdown
input#date[value=""] - mousedown
input#date[value=""] - focusin
input#date[value=""] - pointerup
input#date[value=""] - mouseup
input#date[value=""] - click
input#date[value=""] - keydown
input#date[value=""] - keypress
input#date[value=""] - keyup
input#date[value=""] - keydown
input#date[value=""] - keypress
input#date[value=""] - keyup
input#date[value=""] - keydown
input#date[value=""] - keypress
input#date[value=""] - keyup
input#date[value=""] - keydown
input#date[value=""] - keypress
input#date[value=""] - keyup
input#date[value=""] - keydown
input#date[value=""] - keypress
input#date[value="0001-10-18"] - input
input#date[value="0001-10-18"] - change
input#date[value="0001-10-18"] - keyup
input#date[value="0001-10-18"] - keydown
input#date[value="0001-10-18"] - keypress
input#date[value="0019-10-18"] - input
input#date[value="0019-10-18"] - change
input#date[value="0019-10-18"] - keyup
input#date[value="0019-10-18"] - keydown
input#date[value="0019-10-18"] - keypress
input#date[value="0198-10-18"] - input
input#date[value="0198-10-18"] - change
input#date[value="0198-10-18"] - keyup
input#date[value="0198-10-18"] - keydown
input#date[value="0198-10-18"] - keypress
input#date[value="1988-10-18"] - input
input#date[value="1988-10-18"] - change
input#date[value="1988-10-18"] - keyup
input#date[value="1988-10-18"] - focusout
```

Notice that the `input` and `change` events are fired as soon as I get to the year input.

So what the user types depends on the locale, which we can retrieve if we're given a date object: `date.toLocaleDateString()`.

So what if we accept a `Date` object instead, and then type whatever we get from `date.toLocaleDateString()`? We'll want to make sure to handle all these input/change events too I think.

Reply:
I don't know whether I got it.

[`input#value` (in date types) only accepts `yyyy-mm-dd` format](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/date#Value). I think this might add complexity since we have to format to `yyyy-mm-dd` anyway.

Reply:
As for the `input` and `change` getting fired as soon as we got a valid date, I don't know how to handle that (yet) neither if we need to handle that because we don't know what is the last part of the date to be typed.

If you type last the day (you have typed the year and the month already), the browser will fire the events as soon as you type the day.
If we have a date in format `yyyy-mm-dd` (2020-06-29). We already have typed the year and month, as soon as we type `2`  (from 29):
```
input#date[value=""] - keydown
input#date[value=""] - keypress
input#date[value="2020-06-02"] - input
input#date[value="2020-06-02"] - change
input#date[value="2020-06-02"] - keyup
input#date[value="2020-06-02"] - keydown
input#date[value="2020-06-02"] - keypress
input#date[value="2020-06-29"] - input
input#date[value="2020-06-29"] - change
input#date[value="2020-06-29"] - keyup
```

What do you think?

Reply:
Even though the browser gives the value of an input as `yyyy-mm-dd` format, the *user* will type it in differently and that's the point I was trying to make. We should type it in the same way the user would. Which is why accepting a date and then formatting it with `toLocaleDateString` to know how to type it would probably be best.

Another thing to investigate is how Cypress handles this use case.

Reply:
I think that cypress is supporting only `yyyy-mm-dd`

Reply:
[Cypress only uses `yyyy-mm-dd` format](https://docs.cypress.io/api/commands/type.html#Date-Inputs).

Reply:
That's fine, but what does Cypress do when typing into the input?

I just tested it out with:

```javascript
describe('temp', () => {
  it('fills in dates', () => {
    cy.visit('https://eo909.csb.app/')

    // Comparing what it's like with a regular field first.
    // Using a delay so it's more obvious what the difference is
    cy.findByLabelText(/email/i).type('hello@gmail.com', {delay: 100})

    cy.findByLabelText(/date input/i).type('2020-06-28', {delay: 100})
  })
})
```

![Screen Recording 2020-07-23 at 1 45 16 PM](https://user-images.githubusercontent.com/1500684/88332075-3b26c980-cceb-11ea-8204-3bbf311fc24f.gif)

It looks like they do basically the same thing and don't bother "typing" and instead just set the value. This is surprising to me, but I guess I'm fine with following their lead here.

Unfortunately, for some reason, when running the playground in cypress, it doesn't appear to log events, so I'm not sure what events are being run when cypress conducts its interactions. But I'm guessing we'll want to fire the ones that the user would fire if they typed the value in one character at a time with the exception of change/input events I guess 🤷‍♂️

Reply:
> It looks like they do basically the same thing and don't bother "typing" and instead just set the value. This is surprising to me, but I guess I'm fine with following their lead here.

Are you talking about the `value` format or the events triggered by Cypress?

I think this PR is near to what happens in a browser when the user types into it than Cypress, but we can change the PR if you want.

Thank you.

Reply:
Yeah, I'm thinking this PR is about ready.

I just realized that I merged a PR that results in conflicts with this one. I'm sorry about that. Do you mind fixing those and we can get this merged. Thanks!

Reply:
[I had to move some code around to avoid the `complexity` eslint rule](https://github.com/testing-library/user-event/pull/400/commits/1b34e2400cf3beab455ddf2cd8631324d8a56e75).

Reply:
If that rule ever forces you to make the code *more* complex or confusing then feel free to disable it 😅